Add checkpoint_format to unify all current/future saved weight formats

Qubitium opened the pull request:
Core reason for PR:

Unify the concept of model weight `format` since repo already have coding for 3 weight formats: gptq, marlin, and awq_gemm with 4th pending in #559. Without this PR, existing properties of different weight formats are fragmented and not forward compatible with new formats such as pending #559 which will require another property field. Have one single `format` field to unify them all.

Changes:

* MAJOR `format` unifies `is_marlin_format`,  `awq_qemm`, and pending `gptq_v2` format in pr #559
* `format` defaults to `gptq`
* MAJOR `gptq_method` remapped to `method` 
* MAJOR compat: pinned transformers to < 4.39.0 due to #604
* fix _validate_marlin_device_support() returns false for ada/hopper. Fix #616 
* fix  json arg normalization comparisons did not use lowercased keys and string values
* block awq gemm from participating in quantization: inference only
* added compat for some hf models using `quant_method=marlin` for marlin models using other gptq quantizers
* load quant `method` into  BaseQuantizeConfig. Previously this was ignored, not read from json.
* auto set `use_marlin` to True when model weight format is marlin.
* renamed marlin cache file from  `autogptq_model.safetensors` to more explicit  `autogptq_model_marlin.safetensors`


Reason for the naming changes. `format` means checkpoint format and `method` means quant method but as both properties are specific to `BaseQuantizeConfig`, there is no need to append quant to method and similarly `format` is clear and non-conflicting. 

Backward Compatibility

* autogptq loading pre-pr quantized models marlin/non-marlin have been tested 
* third party using new quantized models will need compat update: i.e. vllm/sglang. 

This is breaking change for downstream users of models quantized by autogptq that do not use the autogptq inference engine.  To mitigate impact, I will parallel pr compat changes to vllm/sglang repos once this is ready to merge.  

TEST Results:

[PASS] Pre-PR: Both marlin and non-marlin load 
[PASS] PR: Both marlin and non-marlin load 

[PASS] test_quantizataion.py 
[PASS] test_serialization.py 
[PASS] test_q4.py
[PASS] test_shared_loading.py
[PASS] test_awq_compatibility_generation.py

Failed tests but these tests also failed on main:
[FAIL] test_triton.py
[FAIL] test_peft.py
[FAIL] test_repacking.py

Qubitium commented:
@fxmarty @PanQiWei  On second review of the PR something did not pass the smell test: awq_gemm. Why is awq_gemm quantization in autogptq? Aren't they different quant methods?

I need clarification on this as I am unsure relevant awq code should be removed entirely. 

In this PR:
1. quantize using `format=awq_gemm` on Yi 9B works
2. loading the quantized `format=awq_gemm` using autogptq fails:

```
 INFO - The layer lm_head is not quantized.
INFO - Loading an AWQ model. This requires repacking the weights, and no repacking cached weight was found. Grab a coffee!
  0%|                                              | 0/224 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/root/python/quantx/gptq_test.py", line 15, in <module>
    model_q = AutoGPTQForCausalLM.from_quantized(model_path, device="cuda:0", use_safetensors=True)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/auto.py", line 140, in from_quantized
    return quant_func(
           ^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/_base.py", line 1185, in from_quantized
    unpacked_qweight, unpacked_qzeros = unpack_awq(
                                        ^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/_utils.py", line 594, in unpack_awq
    qdq_weight_T = weight * scale_mat - scale_zeros_mat.half()
                   ~~~~~~~^~~~~~~~~~~
RuntimeError: The size of tensor a (32768) must match the size of tensor b (4096) at non-singleton dimension 1
```

Qubitium commented:
@fxmarty  based on https://github.com/AutoGPTQ/AutoGPTQ/pull/484 You added awq_gemm load/inference code which are compatible with the autogptq kernels. My question is:

1. Packing of awq_gemm is supported so quantize to format=awq_gemm is fully supported? My previous message shows quantize success but load failed. Unsure if failure is due to packing or loading.

Based on feedback I may need to refractor code as it appears autogptq supports two quant methods "gptq" and "awq" and I need to move "gemm" away from gptq method.

Qubitium commented:
@casper-hansen  Pulling you into the conversation since I am also prepping code to the create same unifying quant method/format PR into autoawq as well, once everything is agreed upon. It will be good if the two major autoquant repos share a compatible quant method/format language spec. This is really helpful downstream when vllm/sglang repos load quants and need to map quant to framework on json load. 

Let's all agree on the naming. Currently my proposal in this PR replaces/unifies:

AutoGPTQ:
* quant_method => method
* is_marlin_format => format
* awq_gemm_checkpoint => format

AutoAWQ:
* quant_method => method (no self-identifying in json)
* version => format

This will help vllm/sglang cleanup qunat loading code moving forward.

casper-hansen commented:
@Qubitium I would prefer we call it version rather than format since that’s already implemented in AutoAWQ. That means no breaking change is required, which we just went through in 0.2.0. This is also something that would have to be phased out over months of time with plenty of warnings, if it were to be considered.

Other than that, I don’t see a reason to rename quant_method to method since that would be another breaking change.

Qubitium commented:
@casper-hansen  Fair. I see how this is more painful to autoawq since the need does not exceed the breaking compat benefits. 

I see this is almost necessary for autogptq going forward though since keeping adding properties is not the answer and since this is a breaking change, might as well cram two breaking changes into one migration step. 

@fxmarty  since `format` and `quant_method` are two breaking changes  merged into one, the impact is still one breaking change. I can parallel pr to all the major downsteam inference engines such as vllm/sglang, etc. to mitigate the impact. 

While I am breaking things, might as well think about adding more things that will make quality of life easier for downstream users. 

How about a tooling version into quant_config? such as `quant_config.producer == "autogptq:0.0.31"` as example. This would help with debugging future quants. Everyone quants but most of time we don't even know which tool/version they used.

fxmarty reviewed (commented):
Thanks a lot for the PR, this is very welcome for simplifying the codebase!

> How about a tooling version into quant_config? such as quant_config.producer == "autogptq:0.0.31" as example. This would help with debugging future quants. Everyone quants but most of time we don't even know which tool/version they used.

I agree this would be a nice addition!

Qubitium commented on auto_gptq/utils/marlin_utils.py:
```diff
@@ -31,7 +33,7 @@ def prepare_model_for_marlin_load(
 ):
     # The model (e.g. model.safetensors) is already serialized in the Marlin format, load it directly.
-    if hasattr(quantize_config, "is_marlin_format") and quantize_config.is_marlin_format:
+    if hasattr(quantize_config, FORMAT_FIELD) and quantize_config.format == FORMAT.MARLIN:
```
the quant_config passed here is actually BaseQuantizeConfig in all code paths. So we don't even need the hasattr() check. and compat is taken care of in BaseQuantizeConfig.from_pretrained(). 

I will refractor so BaseQuantizeConfig is moved from modeling/ to quantization/ to avoid circular imports and make the arg types explicitly clear.

Qubitium commented on auto_gptq/utils/marlin_utils.py:
```diff
@@ -93,33 +95,33 @@ def _get_cached_marlin_save_name(model_name_or_path):
     if os.path.isdir(model_name_or_path):
-        return os.path.join(model_name_or_path, "autogptq_model.safetensors")
+        return os.path.join(model_name_or_path, MARLIN_CACHE_FILE)
     else:
         namespace, subfolder = model_name_or_path.split("/")
         assets_path = huggingface_hub.cached_assets_path(
             library_name="auto_gptq", namespace=namespace, subfolder=subfolder
         )
-        return os.path.join(assets_path, "autogptq_model.safetensors")
+        return os.path.join(assets_path, MARLIN_CACHE_FILE)
```
There is actually a hidden bug here. if user manually update the weights, the cache file is not auto-cleared. This would require a md5 type check of the weights so cache can be invalidated but doing md5 on a huge checkpoint still reads the entire checkpoint.

Qubitium commented on tests/test_quantization.py:
```diff
@@ -37,3 +38,13 @@ def test_quantize(self, use_marlin: bool):
             model.save_pretrained(tmpdirname)
 
             model = AutoGPTQForCausalLM.from_quantized(tmpdirname, device="cuda:0", use_marlin=use_marlin)
+
+            # test compat: 1) with simple dict type 2) is_marlin_format
+            old_quantize_config = {
+                "bits": 4,
+                "group_size": 128,
+                "desc_act": False,
+                "is_marlin_format": use_marlin,
+            }
+            model = AutoGPTQForCausalLM.from_quantized(tmpdirname, device="cuda:0", quantize_config=old_quantize_config)
+            assert(isinstance(model.quantize_config, BaseQuantizeConfig))
```
@fxmarty added test to show 1) weak dict quantize_config input with old `is_marlin_format` is supported 2) weak dict input is force converted into strong BaseQuantizeConfig

Qubitium commented on auto_gptq/utils/marlin_utils.py:
```diff
@@ -93,33 +95,33 @@ def _get_cached_marlin_save_name(model_name_or_path):
-    device_capacity = torch.cuda.get_device_capability()
-    return device_capacity[0] == 8 and (device_capacity[1] == 0 or device_capacity[1] == 6)
-
+    return torch.cuda.get_device_capability()[0] >= 8
```
@fxmarty  latest code added separation of check between marlin_supported and marlin_optimized. logging.info added when marlin_optimized is false. I don't think it warrants a warning level.

My understand of following req is that hooper is not fully optimized in Marlin for maximum performance but still perfomant vs ada/ampere since the marlin kernel is optimized for ada/ampere sm cores/cache sizes. Everything ada/amper has hopper does too: just that hopper will be under-utilized. I don't have hopper to test though.

```
NVIDIA GPU with compute capability >= 8.0 (Ampere or Ada, Marlin is not yet optimized for Hopper)
```

Qubitium commented on auto_gptq/utils/marlin_utils.py:
```diff
@@ -93,33 +95,33 @@ def _get_cached_marlin_save_name(model_name_or_path):
 
 # Adapted from https://github.com/rib-2/marlin/tree/conversion
-def _validate_marlin_compatibility(quantization_config):
+def _validate_marlin_compatibility(quant_cfg):
```
in latest code:  type hint is added so there is no confusion with even more abbrevaited `cfg`.  Very clear and help with pep guidelines due to long strings here.  I would think type hint would fix any confusion on what the var is.

```
def _validate_marlin_compatibility(cfg: BaseQuantizeConfig):
```

Qubitium commented:
@fxmarty Ready for re-review but not merge since test is passing due to failing tests as result of #610. It passed before I rebased to main. Once that issue is fixed I will rebase and re-test. 

Changes since review:

1. `quant_method` name retained. 
2. `checkpoint_format` name used.
3. fixed quantize_config pass as dict (weak) to `from_quantized()`: internal code path is now full strongly typed BaseQuantizeConfig
4. `checkpoint_format` hint arg added to `from_quantized()`: will error if same arg is passed in from both json and `from_quantized()`
5. `marlin_utils` type hint of `BaseQuantizeConfig` added so var `cfg` abbrev is clear
6. separatation of `marlin_compatible` and `marlin_optimized` in cuda device marlin detection: log.info if used but not optimized(*) (hopper)
7. `BaseQuantizeConfig` and quant data types/values are moved from `/modeling/_base.py` to `quantization/config.py`
8. `test_quantization` modified so it will test for both weak dict passing of quantize_config and `is_marlin_format` compat

The remaining issue not fixed from review is diff in intepretation of dasl/marlin note of hopper not optimized. My interpretation is that hopper is suboptimal for marlin in relation to hopper's full capability and not in relation to ada/ampere: it will be just as fast as ada/ampere and but may not be as fast as what hopper is capable of. Hopper has everything Ampere/Ada has with even more sm cores/cache so I don't see how it can be slow. Unfortuantely I have don't have Hopper to test. Can you test this with hf resources?

https://github.com/IST-DASLab/marlin?tab=readme-ov-file#requirements

```
NVIDIA GPU with compute capability >= 8.0 (Ampere or Ada, Marlin is not yet optimized for Hopper)
```

Qubitium commented:
@fxmarty  rebased to main and now all relevant tests are passing (minus the broken ones that never passed even on main)

fxmarty commented on auto_gptq/modeling/_base.py:
```diff
@@ -1082,7 +976,7 @@ def skip(*args, **kwargs):
 
             # TODO: move this logic in an awq_utils.py file.
-            if quantize_config.awq_gemm_checkpoint:
+            if quantize_config.checkpoint_format == CHECKPOINT_FORMAT.GEMM:
```
Could we rename that to `CHECKPOINT_FORMAT.AWQ_GEMM`?

fxmarty commented on auto_gptq/quantization/__init__.py:
```diff
@@ -1,2 +1,3 @@
+from .config import *
```
Can you import here specifically the classes/functions/vars you'd like to expose? I think it is best to avoid star imports.

fxmarty commented on auto_gptq/quantization/config.py:
```diff
@@ -0,0 +1,222 @@
+            raise ValueError(f"Unsupported quantization method: {self.quant_method}")
+
+        if self.checkpoint_format not in valid_checkpoint_formats:
+            raise ValueError(f"Unsupported quantization method: {self.quant_method}, checkpoint format: {self.checkpoint_format}")
```
```suggestion
            raise ValueError(f"The checkpoint format used is {self.checkpoint_format}, and the quantization method is {self.quant_method}. This is not supported, please open an issue at https://github.com/AutoGPTQ/AutoGPTQ/issues.")
```

fxmarty commented on auto_gptq/quantization/config.py:
```diff
@@ -0,0 +1,222 @@
+        CHECKPOINT_FORMAT.GPTQ,
+        CHECKPOINT_FORMAT.MARLIN,
+    },
+    QUANT_METHOD.AWQ: {
+        CHECKPOINT_FORMAT.GEMM
+    }
```
I am wondering why `CHECKPOINT_FORMAT.GPTQ` is not there? I recall the implem I did for AWQ was AWQ GEMM -> gptq format. Does that not work?

Qubitium replied:
Never tested that combo. Did even know it worked. Will test and change pending test result.

Qubitium replied:
@fxmarty  We tested the loading pure awq-gemm model quantized with autoawq on autogptq tip/main. This worked. 

But loading a  Yi6B model quantized by autogptq with groupsize=128,damp=0.005 failed when using setting `awq_gemm_checkpoint=True`

```python
INFO - Loading an AWQ model. This requires repacking the weights, and no repacking cached weight was found. Grab a coffee!
  0%|                                                            | 0/224 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/root/quantx/gptq_test.py", line 21, in <module>
    model_q = AutoGPTQForCausalLM.from_quantized(model_path, use_safetensors=True,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/auto.py", line 140, in from_quantized
    return quant_func(
           ^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/_base.py", line 1173, in from_quantized
    unpacked_qweight, unpacked_qzeros = unpack_awq(
                                        ^^^^^^^^^^^
  File "/root/AutoGPTQ/auto_gptq/modeling/_utils.py", line 603, in unpack_awq
    qdq_weight_T = weight * scale_mat - scale_zeros_mat.half()
                   ~~~~~~~^~~~~~~~~~~
RuntimeError: The size of tensor a (32768) must match the size of tensor b (4096) at non-singleton dimension 1
```

fxmarty replied:
@Qubitium Then let it be. I can have a look later.

fxmarty commented on auto_gptq/quantization/config.py:
```diff
@@ -0,0 +1,222 @@
+
+        if "sym" not in normalized:
+            logger.warning(
+                "The quantization configuration does not contain an entry `sym` (symmetric quantization). This may result in silent errors."
```
```suggestion
                "The quantization configuration does not contain an entry `sym` (symmetric quantization). This may result in silent errors. Defaulting to `sym=True`."
```

fxmarty reviewed (commented):
Thanks a lot, it is very clean. I left a few minor comments and we should be good to go!

Qubitium commented:
[WIP] Not ready for merge/rev-review. AWQ cache tests is failing. 

Changes since last review:

* MAJOR quant cache file moved from static to dynamic: based on quant_method + checkpoint_format. cache file generation moved to `BaseQuantizeConfig.get_cache_file()`. Now marlin and awq cache codes are unified. 
* updated requested error/warning strings and `GEMM`  => `AWQ_GEMM` var name 

After testing, this PR also fixed a bug on tip/main where cache file collision happened between marlin/awq code paths. If I set `use_marlin=True` to True and then switch to `awq_gemm_checkpoint=True`,  the awq code path (and marlin vice versa) would think it has a good cache and load fail. So it is very important the cache files names are pinned to `quant_method` and `checkpoint_format`. 

As note before, there is still a hidden bug in the cache system where if user manually update the weights, the cache files are not invalidated. Personally, I think the caching should be done outside the scope of autogptq as cache management  is cumbersome and prone to state errors as mentioned.

fxmarty reviewed (approved):
LGTM, thanks a lot for this!
